pkg/storage/tsdb/backend/s3/config.go

```diff
@@ -1,21 +1,58 @@
 package s3
 
-import "flag"
+import (
+	"flag"
+	"strings"
+)
 
 // Config holds the config options for an S3 backend
 type Config struct {
-	Endpoint        string `yaml:"endpoint"`
-	BucketName      string `yaml:"bucket_name"`
-	SecretAccessKey string `yaml:"secret_access_key"`
-	AccessKeyID     string `yaml:"access_key_id"`
-	Insecure        bool   `yaml:"insecure"`
+	Endpoints       EndpointFlags `yaml:"endpoints"`
+	SecretAccessKey string        `yaml:"secret_access_key"`
+	AccessKeyID     string        `yaml:"access_key_id"`
+	Insecure        bool          `yaml:"insecure"`
+}
+
+// EndpointFlags is a list of flags for the s3 endpoints
+type EndpointFlags []string
+
+func (e *EndpointFlags) String() string {
+	return "list of endpoint names with a bucket prefix and a date to start using the endpoint from"
+}
+
+// Set does something
+func (e *EndpointFlags) Set(value string) error {
+	// flag.Parse gets called twice, dedupe
+	for _, s := range *e {
+		if s == value {
+			return nil
+		}
+	}
+
+	*e = append(*e, value)
+	return nil
 }
 
 // RegisterFlags registers the flags for TSDB s3 storage
 func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.StringVar(&cfg.AccessKeyID, "experimental.tsdb.s3.access-key-id", "", "S3 access key ID")
 	f.StringVar(&cfg.SecretAccessKey, "experimental.tsdb.s3.secret-access-key", "", "S3 secret access key")
-	f.StringVar(&cfg.BucketName, "experimental.tsdb.s3.bucket-name", "", "S3 bucket name")
-	f.StringVar(&cfg.Endpoint, "experimental.tsdb.s3.endpoint", "", "S3 endpoint without schema")
+	f.Var(&cfg.Endpoints, "experimental.tsdb.s3.endpoint-from", "endpoint names with a bucket prefix and a date to start using the endpoint from (ex: bucket.endpoint|2019-01-01)")
```

Comment:
Few things:
1. The yaml config option is called `endpoints`, while here it's called `endpoint-from`. We should keep consistency
2. It's S3 specific, while we should move this endpoint config to the generic TSDB config and then use it both in S3 and GCS config
3. Have you considered renaming `endpoints` into `buckets`? From a user perspective, may look easier to understand
4. The overall config should work in single bucket mode (no date). I would also suggest to add unit tests to cover both cases
5. I've mixed feelings the way the bucket name and endpoint have been merged. For example, if the bucket name contains dots (ie. it was used in my previous company) it will break. A safer way to handle it would be through a dedicated config block (ie. `BucketConfig`) which contains `bucket_name`, `endpoint` and `starting_date`, and the `buckets` is a list of `BucketConfig`. CLI flags may support only the case of the single bucket, and enforce the use of the YAML config for multi buckets setup (similar to Cortex schema config, which CLI flags support only some legacy use cases).